compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp

```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
-  for (auto op : slice) {
-    rewriter.moveOpBefore(op, insertionPoint);
+  AffineMap interchangeMap =
+      AffineMap::getPermutationMap(interchangeVector, rewriter.getContext());
+  for (auto &map : indexingMaps) {
+    if (!map.isEmpty()) {
+      map = map.compose(interchangeMap);
+    }
   }
   return success();
 }
 
-/// On finding this pattern
-/// ```
-/// %0 = linalg.matmul ins(%arg0, %arg1)
-/// %1 = linalg.matmul ins(%arg0, %arg2)
-/// %2 = linalg.matmul ins(%arg0, %arg3)
-/// ```
-///
-/// where all matmul share an operand can be combined into
-/// rewrite to
-///
-/// ```
-/// %arg1_r = tensor.expand_shape %arg1 [[0, 1], ...] : tensor<?x?xf32> to
-///     tensor<1x?x?xf32>
-/// %arg2_r = tensor.expand_shape %arg2 [[0, 1], ...] : tensor<?x?xf32> to
-///     tensor<1x?x?xf32>
-/// %arg3_r = tensor.expand_shape %arg3 [[0, 1], ...] : tensor<?x?xf32> to
-///     tensor<1x?x?xf32>
-/// %rhs = tensor.concat(%arg1_r, %arg2_r, %arg3_r)
-/// %fused = linalg.generic {
-///     indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d1, d3)>,
-///                      affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
-///                      affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>}],
-///     iterator_types = ["parallel", "parallel", "parallel", "reduction"]}
-///   ins(%arg0, %rhs) ... { ... }
-/// %0 = tensor.extract_slice %fused [0, 0, 0] ... : tensor<1x?x?xf32> to
-///     tensor<?x?xf32>
-/// %1 = tensor.extract_slice %fused [1, 0, 0] ... : tensor<1x?x?xf32> to
-///     tensor<?x?xf32>
-/// %2 = tensor.extract_slice %fused [2, 0, 0] ... : tensor<1x?x?xf32> to
-///     tensor<?x?xf32>
-/// ```
-///
-/// Also accounts for quantized cases where inputs are at lower precision and
-/// accumulate is in higher-precision with truncate getting back to the
-/// quantized sizes.
-static LogicalResult fuseGroup(RewriterBase &rewriter,
-                               HorizontalFusionGroup &fusionGroup,
-                               DominanceInfo &dominanceInfo) {
-  linalg::LinalgOp baseContractOp = fusionGroup.contractionOps.front();
-  Location loc = baseContractOp.getLoc();
-  OpBuilder::InsertionGuard g(rewriter);
-  rewriter.setInsertionPoint(baseContractOp);
-
-  if (failed(moveOperandDefs(
-          rewriter, ArrayRef<linalg::LinalgOp>(fusionGroup.contractionOps),
-          baseContractOp, dominanceInfo))) {
-    return baseContractOp.emitOpError("failed to re-order operand definitions");
+/// Generate the horizontally fused operation as an operation with multiple
+/// results, corresponding to the results of the fused operations. It is assumed
+/// that the LHS of the contraction operations fused horizontally is the same
+/// and have the same indexing map for all the operations. The RHS/outputs of
+/// the operations can be different.
+static std::optional<linalg::GenericOp>
+fuseContractionsHorizontally(RewriterBase &rewriter, Location loc,
+                             MutableArrayRef<Operation *> linalgOps) {
```

Comment:
The returned value is not used at all. If the intention is dumping it for debugging, I think you can log it with `LLVM_DEBUG` within the function. I.e., we can make it return void.

The other question is that why is `linalgOps` mutable? Is it because that they will be replaced with the new results at the end?

Reply:
I think it is good practice to return the newly created op from these kind of transformations. The caller can decide to drop it or not. So its kind of a "norm" in my mind. 
The reason for `MutableArrayRef<Operation *>` was that calling some methods would give me a compile-time error since `ArrayRef<Operation *>` somehow makes all methods go to `const ` method of the underlying object.